src/Game.hs

```diff
@@ -10,69 +11,47 @@ import Data.List
 
 -- | Main "game loop".
 playGame :: IO ()
-playGame = askAboutDimensions
-
--- First success state for playing a game: choose the width for the board size.
-askAboutDimensions :: IO ()
-askAboutDimensions = do
+playGame = do
     putStrLn "Let's choose a board size."
     width  <- getPositiveNumber "How many spots wide should the board be (Q to abort)? "
-    askAboutHeight width
-
--- Second success state: choose the height.
-askAboutHeight :: Maybe Int -> IO ()
-askAboutHeight Nothing = abortGame
-askAboutHeight width   = do
     height <- getPositiveNumber "How many spots high should the board be (Q to abort)? "
-    verifySaneBoard width height
-
--- Third success state is to validate width and height.
-verifySaneBoard :: Maybe Int -> Maybe Int -> IO ()
-verifySaneBoard Nothing _ = abortGame
-verifySaneBoard _ Nothing = abortGame
-verifySaneBoard (Just width) (Just height)
-    | width * height >= 4 = constructBoard width height
-    | otherwise           = do
-        putStrLn "These dimensions make for an uninteresting game.  Try again."
-        putStrLn ""
-        askAboutDimensions
+    let dimens = Dimensions <$> width <*> height
+    case dimens of
+      Just d | isSaneBoard d -> do
+        resetScreen
+        board <- newBoard d
+        playBoard board
+      Just _  -> putStrLn "These dimensions make for an uninteresting game.  Try again.\n"
+      Nothing -> abortGame
 
--- Fourth success state is to construct a board based on the given dimensions.
-constructBoard :: Int -> Int -> IO ()
-constructBoard width height = do
-    putStrLn ""
-    board <- newBoard $ Dimensions width height
-    clearScreen
-    setCursorPosition 0 0
-    playBoard board
+isSaneBoard :: Dimensions -> Bool
+isSaneBoard (Dimensions width height) = width * height >= 4
 
 -- Easy state transition for aborting out of the game.
 abortGame :: IO ()
-abortGame = do
-    putStrLn "Aborting game."
-    putStrLn ""
+abortGame = putStrLn "Aborting game.\n"
 
 -- Fifth success state is to show the board and prompt the player for a move.
 playBoard :: Board -> IO ()
 playBoard board = do
-    putStrLn $ show board
+    putStrLn (show board)
     number <- getPositiveNumber "Enter the number to move (Q to quit): "
-    trySlidingNumber number board
+    case number of
+      Just n  -> trySlidingNumber n board
```

Comment:
That's a good point! Since `slideNumber` already takes care of errors (including negative inputs), you could just allow any Int in `getPositiveInteger` (which probably wouldn't be called get _positive_ integer in that scenario), and then catch the error in `trySlidingNumber`.